### PR TITLE
Kubernetes common

### PR DESCRIPTION
**Pull request by joaomilho**
## Hey, I just made a Pull Request!

This PR introduces the package `@backstage/kubernetes-common`. It contains types to be shared by `@backstage/plugin-kubernetes` and `@backstage/plugin-kubernetes-backend`.

Currently, @backstage/plugin-kubernetes depends on the backend package which in turn depends on the backend-common lib, so if an installation is front end only it suddenly needs the database and kubernetes to be configured.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/kubernetes-common/package.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "@backstage/kubernetes-common",
```
Comment (Fox32) [inline packages/kubernetes-common/package.json:2]: Based on https://backstage.io/docs/architecture-decisions/adrs-adr011, this should actually be in `plugins/kubernetes-common` and the package should have the name `@backstage/plugin-kubernetes-common`?

File: packages/kubernetes-common/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  ): Promise<FetchResponseWrapper>;
+}
+
+// Used to locate which cluster(s) a service is running on
+export interface KubernetesServiceLocator {
+  getClustersByServiceId(serviceId: string): Promise<ClusterDetails[]>;
+}
+
+// Used to load cluster details from different sources
+export interface KubernetesClustersSupplier {
```
Comment (Rugvip) [inline packages/kubernetes-common/src/types.ts:141]: Would say it's best to keep backend-only types like this one in the backend package

File: packages/kubernetes-common/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+export interface KubernetesFetchError {
+  errorType: KubernetesErrorTypes;
+  statusCode?: number;
+  resourcePath?: string;
+}
+
+export interface ConfigClusterLocatorMethod {
+  /**
+   * @visibility frontend
+   */
```
Comment (Rugvip) [inline packages/kubernetes-common/src/types.ts:160]: Noticed these here which seems to be a trick to re-use types between config and the backend implementation. It totally doesn't work for published packages though, and will lead to some bad habits when reading values from config. Seeing as this PR is gonna break that bit even more I think we could take care of it here :grin:

File: packages/kubernetes-common/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# @backstage/techdocs-common
```
Comment (Rugvip) [inline packages/kubernetes-common/CHANGELOG.md:1]: 👀

File: packages/kubernetes-common/package.json (hunk 1 above)
Comment (joaomilho) [reply packages/kubernetes-common/package.json:2]: That makes a _lot_ of sense. It was weird to have a "kubernetes-common" given it's just about the plugin, not kube in general.

File: packages/kubernetes-common/CHANGELOG.md (hunk 4 above)
Comment (joaomilho) [reply packages/kubernetes-common/CHANGELOG.md:1]: copy pasta :D

File: packages/kubernetes-common/src/types.ts (hunk 3 above)
Comment (joaomilho) [reply packages/kubernetes-common/src/types.ts:160]: Oh, gotcha.

Comment (joaomilho) [reply packages/kubernetes-common/src/types.ts:160]: oh, actually... now I'm a bit confused.

The schema.d.ts is in plugins/kubernetes and it refers to ClusterLocatorMethod (which is either ConfigClusterLocatorMethod or GKEClusterLocatorMethod). First, of all, does it make sense to have this config in the frontend package? I don't see this being used there at all.

I feel I should move the shema.d.ts to the backend and all associated types and export 'em from there (all the deps of `Config`). WDYS?

Comment (joaomilho) [reply packages/kubernetes-common/src/types.ts:160]: It's on both actually.

Comment (joaomilho) [reply packages/kubernetes-common/src/types.ts]: So I spoke with Ben & Matthew and added a commit removing the schema from kubernetes. Only the types that are actually shared are now in common, the ones used only in the backend rest there. Type surgery.

**Comment by freben** [comment]
unfortunately a conflict has crept in here, needs rebase

File: plugins/kubernetes-backend/src/types/types.ts [hunk 5]
```diff
@@ -14,145 +14,22 @@
...
-  serviceId: string;
-  clusterDetails: ClusterDetails;
-  objectTypesToFetch: Set<KubernetesObjectTypes>;
-  labelSelector: string;
-  customResources: CustomResource[];
-}
+export type {
+  KubernetesErrorTypes,
+  KubernetesRequestBody,
+} from '@backstage/plugin-kubernetes-common';
```
Comment (freben) [inline plugins/kubernetes-backend/src/types/types.ts:25]: no reason to re-export, it'll just be confusing - individual files that need these types can import from common directly instead

File: plugins/kubernetes-backend/src/types/types.ts [hunk 6]
```diff
@@ -14,145 +14,22 @@
...
-    params: ObjectFetchParams,
-  ): Promise<FetchResponseWrapper>;
-}
-
-// Used to locate which cluster(s) a service is running on
-export interface KubernetesServiceLocator {
-  getClustersByServiceId(serviceId: string): Promise<ClusterDetails[]>;
-}
+export type KubernetesFetchError = KubernetesFetchErrorCommon;
+export type FetchResponse = FetchResponseCommon;
```
Comment (freben) [inline plugins/kubernetes-backend/src/types/types.ts:28]: What's the point of these?

File: plugins/kubernetes-common/package.json [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    "url": "https://github.com/backstage/backstage/issues"
+  },
+  "dependencies": {
+    "@backstage/catalog-model": "^0.7.6",
+    "@kubernetes/client-node": "^0.14.0"
+  },
+  "devDependencies": {
+    "@backstage/cli": "^0.6.5"
+  },
+  "jest": {
```
Comment (freben) [inline plugins/kubernetes-common/package.json:44]: remove this one

File: plugins/kubernetes-backend/src/types/types.ts (hunk 5 above)
Comment (joaomilho) [reply plugins/kubernetes-backend/src/types/types.ts:25]: Sure, wanted to minimize the change, bit it's best indeed.

File: plugins/kubernetes-common/package.json [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    "types": "dist/index.d.ts"
+  },
+  "homepage": "https://backstage.io",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/backstage/backstage",
+    "directory": "plugin/kubernetes-common"
+  },
+  "keywords": [
+    "techdocs",
```
Comment (freben) [inline plugins/kubernetes-common/package.json:20]: not really :)

File: plugins/kubernetes-common/package.json (hunk 7 above)
Comment (freben) [reply plugins/kubernetes-common/package.json:44]: still here!

**Review by freben — APPROVED** [review]
Tiny nits left, let's merge

**Comment by tragiclifestories** [comment]
> so if an installation is front end only it suddenly needs the database and kubernetes to be configured.

Damn, I was wondering why that was. Thanks! 🙌

File: plugins/kubernetes-common/package.json (hunk 8 above)
Comment (joaomilho) [reply plugins/kubernetes-common/package.json:20]: ha, sorry

**Pull request merged**